sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -130,6 +131,11 @@ class MicroBatchExecution(
 
   protected var watermarkTracker: WatermarkTracker = _
 
+  // Store checkpointIDs for state store checkpoints to be committed or have been committed to
+  // the commit log.
+  // operatorID -> (partitionID -> uniqueID)
+  private val currentCheckpointUniqueId = MutableMap[Long, Array[String]]()
```

Comment:
Maybe this is better to be put into the stream execution context

Reply:
`operatorID -> (partitionID -> uniqueID)`, is this supposed to mean a map of maps? If so, then why is the type of `currentCheckpointUniqueId` just a single map?

I also don't fully understand why we would need a unique map for every operator X partition. Why is it not sufficient to have the following protocol, where we have _one_ unique ID for every batch:

For the first batch, an ID is created and sent to all executors. When all tasks finish, that ID is persisted into the commit log. It is also kept in memory for the subsequent batch.

For any other batch, if there does not exist an ID in memory from the previous batch, then it must be read from the commit log and brought into memory. (This is the restart case.)

Then, using the ID in memory from the previous batch (call that `prevId`), this is sent to all executors in the physical plan, as well as a new ID for the current batch (call this `currId`). Before any processing start, executors  must load and use the state for `prevId` to process the current batch. Then, they can start processing, and they upload their state as `<state file name>_currId.<changelog|snapshot>`.

What's wrong with that?

Reply:
Right now, the uniqueID is generated in executor. As a potential optimization, the driver can send a uniqueID to all executors, but executors still need to modify it to make it unique among all attempts of the same task. After doing that, the IDs won't be unique anymore, so we need different IDs per partition.